pkg/operator/kit/resource.go

```diff
@@ -0,0 +1,158 @@
+	MaxRetries int
+
+	// The host where the Kubernetes master is found.
+	MasterHost string
+
+	// An http connection to the Kubernetes API
+	KubeHTTPCli *http.Client
+}
+
+// CustomResource is for creating a Kubernetes TPR/CRD
+type CustomResource struct {
+	// Name of the custom resource
+	Name string
+
+	// Version which should be defined in a const above
+	Version string
+
+	// Description that is human readable
+	Description string
+}
+
+// CreateCustomResources creates the given custom resources and waits for them to initialize
+func CreateCustomResources(context KubeContext, resources []CustomResource, group string) error {
+	for _, resource := range resources {
+		if err := CreateCustomResource(context, resource, group); err != nil {
+			return fmt.Errorf("failed to init resource %s. %+v", resource.Name, err)
+		}
+	}
+
+	for _, resource := range resources {
+		if err := waitForCustomResourceInit(context, resource, group); err != nil {
+			return fmt.Errorf("failed to complete init %s. %+v", resource.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// CreateCustomResource creates a single custom resource, but does not wait for it to initialize
+func CreateCustomResource(context KubeContext, resource CustomResource, group string) error {
+	logger.Infof("creating %s resource", resource.Name)
+	r := &v1beta1.ThirdPartyResource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s.%s", resource.Name, group),
+		},
+		Versions: []v1beta1.APIVersion{
+			{Name: resource.Version},
+		},
+		Description: resource.Description,
+	}
+	_, err := context.Clientset.ExtensionsV1beta1().ThirdPartyResources().Create(r)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create %s resource. %+v", resource.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func waitForCustomResourceInit(context KubeContext, scheme CustomResource, group string) error {
+	restcli := context.Clientset.CoreV1().RESTClient()
+	uri := resourceURI(scheme, "", group)
+	return Retry(context, func() (bool, error) {
+		_, err := restcli.Get().RequestURI(uri).DoRaw()
+		if err != nil {
+			logger.Infof("did not yet find resource %s at %s. %+v", scheme.Name, uri, err)
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	})
+}
+
+func watchResource(context KubeContext, scheme CustomResource, namespace, resourceVersion, group string) (*http.Response, error) {
+	uri := fmt.Sprintf("%s/%s?watch=true&resourceVersion=%s", context.MasterHost, resourceURI(scheme, namespace, group), resourceVersion)
+	logger.Debugf("watching resource: %s", uri)
+	return context.KubeHTTPCli.Get(uri)
+}
+
+func resourceURI(scheme CustomResource, namespace, group string) string {
+	if namespace == "" {
+		// creates a uri that is for retrieving or watching a resource in all namespaces. For example:
+		//   /apis/rook.io/v1alpha1/clusters
+		return fmt.Sprintf("apis/%s/%s/%ss", group, scheme.Version, scheme.Name)
```

Comment:
ah, here's one of those automatic pluralizations that the CRD blog post was talking about improving